src/sources/geniuscom.cpp

```diff
@@ -71,57 +75,112 @@ std::vector<LyricDataRaw> GeniusComSource::search(const LyricSearchParams& param
+            cJSON_Delete(json);
+            return {};
+        }
+
+        int results = 0;
+        const cJSON* search_hit = nullptr;
+        cJSON_ArrayForEach(search_hit, search_hits) {
+            results++;
+            if (results > RESULT_LIMIT) {
+                break;
             }
-            break;
+
+			const cJSON* search_result = cJSON_GetObjectItem(search_hit, "result");
+			const cJSON* search_path = cJSON_GetObjectItem(search_result, "api_path");
+			const cJSON* search_title = cJSON_GetObjectItem(search_result, "title");
+			const cJSON* search_artist = cJSON_GetObjectItem(search_result, "artist_names");
+
+			LyricDataRaw result = {};
+			result.source_id = id();
+			result.lookup_id = search_path->valuestring;
+            result.source_path = search_path->valuestring;
+            result.title = search_title->valuestring;
+            result.artist = search_artist->valuestring;
+
+			song_metadata.push_back(result);
         }
+
+        cJSON_Delete(json);
     }
 
-    if(lyric_text.empty())
+    return song_metadata;
+}
+
+bool GeniusComSource::lookup(LyricDataRaw& data, abort_callback& abort)
+{
+    auto request = http_client::get()->create_request("GET");
+    request->add_header(API_KEY_HEADER);
+
+    std::string url = "https://api.genius.com" + std::string(data.lookup_id) + "?text_format=plain";
+
+    LOG_INFO("Page %s retrieved", url.c_str());
+    std::string lyric_text;
+
+    pfc::string8 content;
+    try
     {
-        throw new std::runtime_error("Failed to parse lyrics, the page format may have changed");
+        file_ptr response_file = request->run(url.c_str(), abort);
+        response_file->read_string_raw(content, abort);
+        // NOTE: We're assuming here that the response is encoded in UTF-8 
     }
-    else
+    catch (const std::exception& e)
     {
-        LOG_INFO("Successfully retrieved lyrics from %s", url.c_str());
-        const std::string_view trimmed_text = trim_surrounding_whitespace(lyric_text);
-
-        LyricDataRaw result = {};
-        result.source_id = id();
-        result.source_path = url;
-        result.artist = params.artist;
-        result.album = params.album;
-        result.title = params.title;
-        result.type = LyricType::Unsynced;
-        result.text_bytes = string_to_raw_bytes(trimmed_text);
-        return {std::move(result)};
+        LOG_WARN("Failed to download genius.com page %s: %s", url.c_str(), e.what());
+        return {};
     }
-}
 
-bool GeniusComSource::lookup(LyricDataRaw& /*data*/, abort_callback& /*abort*/)
-{
-    LOG_ERROR("We should never need to do a lookup of the %s source", friendly_name().data());
-    assert(false);
+    LOG_INFO("Successfully retrieved lyrics from %s", url.c_str());
+
+    {   // Parser gets its own scope
+        cJSON* json = cJSON_Parse(content.c_str());
+        if (json == nullptr) {
+            LOG_WARN("Failed to download genius.com page %s: %s", url.c_str(), "Failed to parse JSON");

```

Comment:
We didn't fail to *download* the page, we just failed to *parse* it, in which case we should log the thing that we failed to parse (the musixmatch source is a decent example).